packages/react/use-previous/src/usePrevious.tsx

```diff
@@ -3,15 +3,18 @@ import * as React from 'react';
 function usePrevious<T>(value: T) {
   // The ref object is a generic container whose current property is mutable ...
   // ... and can hold any value, similar to an instance property on a class
-  const ref = React.useRef<T>(value);
+  const ref = React.useRef({ value, previous: value });
 
-  // Store current value in ref
-  React.useEffect(() => {
-    ref.current = value;
-  }, [value]); // Only re-run if value changes
-
-  // Return previous value (happens before update in useEffect above)
-  return ref.current;
+  // We compare values before making an update to ensure that
+  // a change has been made. This ensures the previous value is
+  // persisted correctly between renders.
+  return React.useMemo(() => {
+    if (ref.current.value !== value) {
```

Comment:
Primarily to ensure [consistent behaviour on CSB](https://github.com/codesandbox/codesandbox-client/issues/6357)

Reply:
Why redundant though? If we didn't check, then wouldn't the value be reassigned for any change?

Reply:
In an alternative version we could get away without the check:

https://stackblitz.com/edit/react-ejzqys?file=src/App.js

but it was unreliable in CSB:

https://codesandbox.io/s/loving-lamport-fimun?file=/src/App.js

Which is why the code was adjusted the use the check.

Reply:
Ah yeah I see, thanks for the clarification.
Man that's frustrating to have to add code in our lib to please CSB though…

Reply:
Yeah but there is also this:

> In the future, React may choose to “forget” some previously memoized values and recalculate them on next render, e.g. to free memory for offscreen components. Write your code so that it still works without useMemo — and then add it to optimize performance.
> https://reactjs.org/docs/hooks-reference.html#usememo

So, relying on deps to signify "value has changed" isn't entirely accurate with `useMemo`.